[5.5] Improve the Eloquent push method

sileence opened the pull request:
This is a continuation of https://github.com/laravel/framework/pull/19681 and https://github.com/laravel/framework/pull/16384.

The idea is to persist the inverse relations first but also to create the correct association between the two models in the relation. For example if a `Post` belongs to a `Category` and has many `Comment`. We should:

1. Persist the `Category` first and associate it to the `Post`
2. Save the `Post`
3. Add the foreign key (i.e. `post_id`) to the `Comment` and then save the `Comment`

To achieve that: 

I added a new `isInverse` method to the Relation class, it returns `false` by default and I override it to return `true` from the `BelongsTo` class. 

I also added a `push()` method in the Relation class that I can override in the child classes (like `BelongsTo` and `HasMany`) in order to add the correct association between a child and a parent or a parent and a child model, etc.

Issues so far: 

In order to achieve this I need to create the relation object but not all the relation have it, you can see a quick hack in my solution but of course this is something that should be fixed.

I'm also thinking about maybe moving the push method to a trait or even another class. But I just wanted to create like a first version of this idea to get feedback. What do you think?

sileence committed:
Save relations between relatives automatically when using push in Eloquent

sileence commented:
these are the tests / examples I've been using: 

https://github.com/sileence/eloquent_push_tests/blob/master/tests/Unit/EloquentPushTest.php

I created a new project with L5.5 😊

tillkruss commented:
Tests are failing.

sileence commented:
Thank you, @tillkruss. I fixed the problem. I was trying to simplify this line: https://github.com/laravel/framework/pull/19793/files#diff-d8e24a5815a26f7211e66db787be647fR495 In my mind calling collect while passing a model should allocate the model as the first item in the collection. Maybe this behaviour could be added for the Eloquent collection?

Anyway, so this new version should work in the same way as the old version unless the push method is overwritten in the Relation class. So far I've added the expected functionality to `HasMany` and `BelongsTo`.

All feedback is appreciated.

sileence commented:
I worked hard to avoid adding many extra methods / lines of code to the Model class.

taylorotwell commented:
So this PR solves all the problems you are aware of?

sileence commented:
@taylorotwell so far, yes, these integration tests are passing:

https://github.com/sileence/eloquent_push_tests/blob/master/tests/Unit/EloquentPushTest.php

- has many
- belongs to
- belongs to many.

In order to add extra logic for other relations, we just need to override the `push($model)` in the Relation class. Otherwise it fallbacks to simply calling push in the model (current behaviour).

By the way I'm going to ping @barryvdh as he was also dealing with this issue some time ago :)

sileence commented:
@barryvdh if you have a minute, could you please have a look at this PR and let us know your thoughts? I'd highly appreciate it since you ran into the same issue before and you were also working on a solution for this problem.

jgadrow commented:
I think we can get greater readability in the Model::push() method, cut out the PushableRelation class (which would require keeping it synchronized future-forward with the duplicated logic in the Relation class), and also reduce memory footprint.

This all comes at the cost of running foreach on $this->relations a second time. IMO, a worthy trade.

This being said, I do like the additions of the isInverse and Relation::push as that elegantly handles scenarios with non-core Relations (not applicable for my current needs, but I can see where it would be useful).

Essentially, if the code of Model::push() was written as:

```
{
    return $this->pushInverseRelations()
        && $this->save()
        && $this->pushNonInverseRelations();
}
```

You cut out the local variables needed for the separate pushes (reduced memory footprint), the logic is super-clear (readability gain), and because it's in separate methods you can just do:

```
foreach ($this->relations as $name => $data) {
    if (method_exists($this, $name) && $this->$name()->isInverse()) {
        // Push the relation...
    }
}
```

For non-inverse, the check is inverted. This negates the need for the PushableRelation class due to the check for the existence of the relation method prior to calling it and checking isInverse(). It does the same thing as the PushableRelation class would do... defaults to treating it as a non-inverse relation. Only without the expectation of keeping the two separate classes synchronized.

If we feel the need to keep the PushableRelation class and all that jazz... at least consider refactoring the code in both places to a Trait so that synchronization can be achieved without possessing the "tribal knowledge" of keeping it synchronized.

Hopefully that's clear enough... let me know if it's not! :)

sileence commented:
@jgadrow thank you for your comments. I had something similar before (sadly I squashed the commits already). I didn't like checking for method_exists in 2 different places and the repetition of both push relation methods, so I came with the solution of the kind-of null object "PushableRelation". I think the push routine code would benefit if I could move it to its own class.

sileence committed:
Refactor most of the logic to push model relationships to a new class, delete "pushable relation" because it's no longer necessary.

lucasmichot commented on src/Illuminate/Database/Eloquent/Relations/RelationPusher.php:
```diff
@@ -0,0 +1,71 @@
+        // through this object, since the relation can contain extra logic to
+        // associate relatives, otherwise just push the model(s) directly.
+        return $this->models->filter()->every(function ($model) {
+            return $this->relation ? $this->relation->push($model) : $model->push();
```
So for a collection of 1,000 models, you will check 1,000 times that `$this->relation` is not null, right @sileence 
 ?
Why not just checking only once ?

lucasmichot replied:
This would be way much faster:
```php
$models = $this->models->filter();

if ($this->relation) {
    return $models->every(function ($model) {
        return $this->relation->push($model);
    });
}

return $models->every(function ($model) {
    return $model->push();
});
```

sileence replied:
I did something similar, please check the update.

sileence committed:
Optimise push method

sileence commented:
@lucasmichot you are right! Although I'm not trying to achieve that level of optimisation and I hope no one tries to push 1000 models at the same time... I believe this changes makes more sense (makes the method easier to read). What do you think now?

sileence commented:
Ok, I think I'm done with this proposal.

jgadrow commented:
I don't think the new changes were a step in the right direction. So, you were unhappy with calling method_exists in two places (I get it, very similar checks) but you're happy with an undeclared class dependency caused by "new"-ing up an instance within the body of a method?

The simpler solution for refactoring to a single check would be to make a separate method:

```
protected function relationIsInverse($relation)
{
    return method_exists($this, $name) && $this->$name()->isInverse();
}
```

Could even make that method public as it's just about gathering meta data so it's harmless to expose.

I feel like we're working too hard to use a Collection here when it's not really necessary. We're operating on a resource that can be shared between the two operations ($this->relations). In current code, we are splitting that resource prior to operation. The reason being that each operation doesn't need to worry about  filtering the resource to only the subset it intends to operate upon. This decision is intended to save a little bit of CPU processing. The *cost* of that decision though is, literally, *double* the memory consumption. Looping over an array is an extremely fast operation so it would take many, many relations before it would become a noticeable issue. In fact, by introducing the added abstractions for working with it through a Collection, we might even be *increasing* the required CPU processing! You'd need to run it through a benchmark to be certain.

So, we're trying to optimize CPU usage but are uncertain if it even *is* an optimization... but we are *definitely* less memory efficient because of it. Wouldn't it make sense to just save the memory then and run two foreach loops in separate processing methods? Especially when memory tends to be the more limited resource in PHP development?

sileence commented:
@jgadrow I'm not trying to save a little bit of CPU processing or anything like that. I was just trying to:

- make the code more readable by avoiding nested loops & conditionals
- avoid duplication (by having 2 very similar methods doing almost the same thing)
- adding extra methods to the Eloquent model (as this class is extended by every model in your app, maybe someone wants to define a `pushRelations` method for some reason, I didn't really want to add more methods to Eloquent).

`RelationPusher` is more or less a decorator / wrapper, I don't see why anyone would need to mock/replace that class and if someone needs to, well just override the push method.

jgadrow commented:
Good point about "adding extra methods." If that's the case though, just make them private instead of protected. That way they aren't inherited and can be freely re-defined later without interfering with the operation of the Model::push method.

Your other points, IMO, don't outweigh doubling the memory consumption. Though maybe I'm outnumbered in this as already if I var_dump the "wrong" object in Laravel I get an Out of Memory error. ;)

The reason I bring it up is that even if you assume a simple scenario as you described at the beginning of the PR, you could definitely expect at least 11 objects to be contained in your relation data (ie: a Category and 10 Comments). Each of those objects has their own data and relations.

We're not able to know for certain what kind of memory usage we are talking about but splitting *prior* to processing doubles *whatever* the number is. Even if you have an unreasonably small assumption about storage requirements for these objects:

- 255 chars per comment body (255 bytes)
- integer for user identifier (4 bytes)
- integer for timestamp (4 bytes)
- 1 char for category "name" (1 byte)

You have a "base" consumption of (255+4+4)*10+1 = 2631 bytes (~2.57k)

Your $inverse and $nonInverse variables together store a full copy of the base data meaning you're consuming 5KB of RAM when you need half that much. Again, keep in mind that we are setting our memory assumptions to an *unreasonably small* amount! It's far more likely that the User would be another relation and that the comment body would be a text field capable of storing *far* more data! Possibly even on the order of 2GB *per* comment which then blows up our possible memory usage to 40GB!

Granted, that is also unrealistic... but should illustrate the reason for concern. We have a *much* wider degree of variance in terms of memory than we can expect for processing the $this->relations variable.

I would also argue that (read from the execution entry point of Model::push()... I have organized them by visibility and then alphabetically out of habit lol):

```
private function doPush($inverse = false) {
    foreach ($this->relations as $name => $related) {
        // Skip relations of the type we are not concerned with.
        // This is the only part of the logic which *might* confuse someone... and can be gracefully
        // clarified by a one-line comment.
        if ($inverse !== $this->isInverseRelation($name)) {
            continue;
        }

        // This is where your previous pushRelations method comes in.
        // Only it is focused on pushing only a single relation instead of *all* of them now.
        // This comment is only so I could leave out that method and wouldn't actually be here.
        if (!$this->pushRelated($related)) {
            return false;
        }
    }

    return true;
}

private function isInverseRelation($relation)
{
    return method_exists($this, $relation) && $this->$relation()->isInverse();
}

private function pushInverseRelations()
{
    return $this->doPush(true);
}

private function pushNonInverseRelations()
{
    return $this->doPush();
}

public function push()
{
    return $this->pushInverseRelations()
        && $this->save()
        && $this->pushNonInverseRelations();
}
```

Is brief and clear in intent. The current code has a large comment block just to identify *what* is happening and *why.* The intent of the two push methods is only to increase readability but if you're really against them, you could replace them with the doPush calls directly... though it would immediately make the code less clear upon casual observation.

Yes, there are a few conditionals within a loop... but you have to know, at worst, only 2 levels of info. All of which is immediately available only a few lines up.

This logic means the only copy we hold in memory is for a *single* relation at a time. Not a massive improvement on our minimalist scenario (as the "large" memory duplication is only for the comments relationship) but for *any* scenario where we have multiple "many" relationships, this is a large gain in terms of memory usage.

sileence commented:
@jgadrow please correct me if I'm wrong, when I'm passing the models to the `RelationPusher` I'm not duplicating them because they are passed by reference? The only extra memory my code needs is to create the relation pusher object and inside it the only data that is duplicated is the name of the relationship (a simple string), the rest of the data are objects.

But every time you call `$this->$relation()` you are actually instantiating a new object that perform a series of operations. 

Push is a heavy operation indeed, but I think the main operation is going to be on the database  level, not in php.

jgadrow commented:
his line is where I believe you are creating the copies:
```
list($inverse, $nonInverse) = collect($this->relations)
```

I'd be happy to be wrong... but if you make a change to $this->relations[$nameOfAnInverse] is it instantly reflected in the $inverse data? If so, then no copies are taking place. If nothing changes... then you've made a copy. However, in the event that your related data is not a Collection, then the the line where you do collect([$models]) *is* performing a copy. In order for it to refer to the original you'd have to use the reference operator.

I cannot actually be certain for the other case unless I went and read through the implementation of how Collections work... which is kind of why I say the other method is more clear in what is happening.

It is true that calling the relation method instantiates a new object. However, in my code above I'm not assigning the result of this operation to a variable anywhere which means it is automatically free for garbage collection. In your class you're storing the instance in a property which means it cannot be collected until all references to your class are also freed.

I agree that most of the pressure is going to be on the database and is most likely to be the bottleneck in even the most optimal of scenarios. As such, that's not really a point of consideration as nothing we do here can further optimize that aspect.

sileence commented:
@jgadrow that method split a collection in two, so it duplicates the collection but it keeps the reference to the same objects (models), it doesn't clone the models.

jgadrow commented:
After running a quick test, you're right. It does still pass by reference when performing "[$models]" (could have sworn I had issues with that an ages past...). As such, I'm fine with it now. :)

taylorotwell commented:
What's the status of this? Is it ready to merge? Are there known issues it doesn't solve?

sileence commented:
@taylorotwell yes, it's ready to merge. My integration tests (that hit the DB) and the unit tests are passing and I haven't found any other issue.

bloggervista commented:
@Yes I also checked and as it is passed by reference I am not getting too much memory uses. And unit tests are passing to me as well :)

taylorotwell commented:
I still can't even decide if it feels logical to have "push" traverse upwards a relation graph. Seems counter-intuitive.

sileence commented:
@taylorotwell well another possible solution would be to omit the inverse side of the relationships entirely and just link and persist the children related models.

taylorotwell commented:
@sileence isn't that what push already does?

sileence commented:
@taylorotwell this proposal will save all relations (as Laravel currently does).

The differences / benefits of this proposal is to save *in the right order* and assign the foreign keys.

What I tried to say in my last comment is that  instead of partitioning the relations in non-inverse and inverse ones to save them both another option would be to filter and save only the non-inverse relations?

sileence committed:
Make push method persist non inverse relations only and add pushAll to persist everything

sileence commented:
I changed `push()` to persist the non inverse relations only and I also added `pushAll()` in case the developer wants to persist everything. This is just an idea / proposal.

taylorotwell commented:
I'm going to wait on this for now.

ArondeParon commented:
Still very much interested in this feature. Persisting a new model with it's relations in one-go would be a a really great idea.

TwIsL commented:
Still very interested in implementing this feature. Can we know the status of it?
